Subgraph: rename to align with contracts

0xGabi opened the pull request:
This PR rename the current schema to align more with the smart contracts. Apologies if I introduce further confusion with the naming. The other current options we discussed are: `Container`, `Packet` and `Item`.

Let's use this branch to stick to one of the above.

0xGabi committed:
Subgraph: rename to align with contracts

0xGabi committed:
Merge remote-tracking branch 'origin/master' into subgrpah-rename

0xGabi committed:
Subgraph: rename mapping build methods

0xGabi committed:
Subgraph: fix entity argument name

bpierre commented on packages/govern-subgraph/schema.graphql:
```diff
@@ -63,37 +63,37 @@ type ERC20 @entity {
   id: ID!
   queue: GovernQueue!
-  status: PacketStatus!
+  status: Status!
```
Should we use the term “state” to reflect the contracts?

Also it seems that the hierarchy is the following in the contracts (using TS to describe the relations):

```ts
type Queue = {
  initConfig: Config

  // In the contract only the state is stored in the queue using
  // the container hash as a key, but the subgraph could be more
  // straightforward and have containers in the queue directly.
  queued: Container[]
}

type Config = {
  executionDelay: BigInt
  scheduleDeposit: Collateral
  challengeDeposit: Collateral
  vetoDeposit: Collateral
  resolver: string
  rules: string
}

type Collateral = {
  address: string
  amount: BigInt
}

type Container = {
  config: Config

  // I don’t think it would make sense for a Container to belong to different
  // queues, so I think we can store the state on the Container directly.
  state: ContainerState

  payload: ContainerPayload
}

enum ContainerState {
  None,
  Scheduled,
  Challenged,
  Approved,
  Rejected,
  Cancelled,
  Executed
}

type ContainerPayload = {
  nonce: BigInt
  executionTime: BigInt
  submitter: string
  executor: Executor
  actions: Action[]
  allowFailuresMap: string
  proof: string
}

type Action = {
  to: string
  value: BigInt
  data: string
}
```

Should we try to reproduce this in the subgraph?

Note: I excluded having references to parents for readability, but in the subgraph we could also expose them (e.g. by having `Action.container`).

Evalir replied:
100% agree with this! This is quite similar to the first version of the subgraph, and I think having a structure closer to the contracts lets us understand the subgraph faster.

> Note: I excluded having references to parents for readability, but in the subgraph we could also expose them (e.g. by having Action.container).

Yep, I agree having the parent references in the schema.

kanaagawa replied:
This structure looks really clean and super close to the smart contract. I diverge a bit from that with the intention to make it a bit more straight forward. And probably has the opposite effect 🙈

I think I miss-understood the importance of having a `Container` that have both a reference to the `Payload` and the `Config`. Because there are scenarios where the `Config` changed and you still want to have the `Payload` linked to the previous `Config`.

bpierre commented on packages/govern-subgraph/schema.graphql:
```diff
@@ -130,7 +130,7 @@ type Role @entity {
 }
 
-enum PacketStatus {
+enum Status {
```
Does it make sense to have `Vetoed` as a status? Should we instead set it to `Cancelled` [like in the contract](https://github.com/aragon/govern/blob/master/packages/govern-core/contracts/pipelines/GovernQueue.sol#L203) and, to store the different events, have a `Container.history = ContainerEvent[]` that would contain the entire history of events happening on a given container?

kanaagawa replied:
Yes, that's a good point. I think we previously had a `Vetoed` state but I agree we should stick to the contracts. I like the idea of having a `Container` history 🙌

0xGabi committed:
Merge remote-tracking branch 'origin/master' into subgrpah-rename

0xGabi committed:
Subgraph: add Container entity

Evalir commented on packages/govern-subgraph/schema.graphql:
```diff
@@ -24,15 +24,15 @@ type Govern @entity {
   id: ID!
   address: Bytes!
-  config: Config
+  configs: [Config!]
```
I'd prefer to only keep a single config here—We could totally choose to track the history later, but let's keep it close to the contracts.

kanaagawa replied:
I choose to keep track the history of configs because we need to know the specific configuration that a `Container` was schedule with. Otherwise when we update the config of the `Queue` the `Container` will have a the wrong configuration data.

Evalir commented on packages/govern-subgraph/schema.graphql:
```diff
@@ -78,22 +80,22 @@ type Payload @entity {
+  container: Container!
   collateral: Collateral!
   createdAt: BigInt!
 }
```
This entity, now that we have containers, feels a bit redundant.

Evalir reviewed (commented):
Some more comments, I'm probably gonna do these soon:

Evalir committed:
Align subgraph schema with contracts

Evalir commented on packages/govern-subgraph/schema.graphql:
```diff
@@ -16,28 +16,18 @@ type Govern @entity {
   metadata: Bytes
-  games: [OptimisticGame!] @derivedFrom(field: "executor")
-  executions: [Execution!]
+  correspondingEntries: [RegistryEntry!] @derivedFrom(field: "executor")
```
Still not happy with this name, but going with it for now. Suggestions are welcome. :D

bpierre replied:
If we go for the name `RegistryEntry` to refer to the Dao <> Queue association, should we keep it consistent and use `registryEntries` here?

Or maybe we could use the term `Registration` (`registrations` here) to reflect [the verb used for the method](https://github.com/aragon/govern/blob/master/packages/erc3k/contracts/ERC3000Registry.sol#L18)? We could also do `DaoQueueRegistration` but it starts to be a bit long :smile:

Also I don’t think the array can be null?

```suggestion
  correspondingEntries: [RegistryEntry!]! @derivedFrom(field: "executor")
```

kanaagawa replied:
I think games were pretty appealing given that DAOs are going to be participating on this optimistic games. What is the reason to instead linked to entities on a registry?

Evalir replied:
DAOs might be participating in optimistic games, but they're _not_ optimistic games by themselves. I think it's easier to think of these entities as entries in the registry, mainly because it's what they are in the contracts.

Evalir replied:
Btw @bpierre Yep, def agree about the naming change. What about sticking to `registryEntry` and `registryEntries`? I think it sounds natural, even if it's 100% aligned with the method's verb. :)

bpierre replied:
Sounds good to me! @0xGabi what do you think?

kanaagawa replied:
Yes, sound good

Evalir commented on packages/govern-subgraph/schema.graphql:
```diff
@@ -86,13 +84,8 @@ type Action @entity {
-
+# I honestly think this should not be here,
+# but we can handle this later.
 type Challenge @entity {
```
As I say here, I really think that this entity should not exist here. It seems out of scope for the subgraph itself because if an action gets disputed, it's really up to the dispute resolution protocol to be able to get this info from its own graph; we should only facilitate finding the actual disputed action so that they're able to show context regarding it.

Removing this required some hefty changes from the subgraph, so I avoided it, but definitely I think at most, we should just set it to the actual `Challenged` state, and on the `history` field provide some more info on the `data` field.

Evalir commented on packages/govern-subgraph/schema.graphql:
```diff
@@ -114,15 +107,6 @@ type Evidence @entity {
   createdAt: BigInt!
 }
 
-type Veto @entity {
```
For removing this, on the `data` field of the latest container `history` event that will get recorded I just put the `reason`. That way you should be able to see why this was vetoed.

Evalir reviewed (commented):
Changed some more things, mainly to align with @bpierre's suggestions. This is not 100% there yet, but we're very close! The only thing left to be in sync with the server, should be removing the `Challenge` entity.

bpierre commented on packages/govern-subgraph/schema.graphql:
```diff
@@ -52,48 +42,53 @@ type Config @entity {
+  id: ID!
+  type: ContainerEventType
+  createdAt: BigInt
+  data: [String]!
```
Wondering if we should use `[Bytes!]!` here, to accept any type of data?

Evalir replied:
Yep yep, I was meaning to change this but that slipped by. 😄

bpierre commented on packages/govern-subgraph/schema.graphql:
```diff
@@ -130,13 +116,26 @@ type Role @entity {
+  Execute,
+  Veto,
+  Rule,
+  EvidenceSubmission
```
Maybe we should rename this `SubmitEvidence` to make it a verb like the others?

bpierre commented on packages/govern-subgraph/src/GovernQueue.ts:
```diff
@@ -11,148 +11,182 @@ import {
+  container.state = EXECUTED_STATUS
 
-  packet.save()
+  container.save()
```
Should we create a `containerEvent.type = EXECUTE_CONTAINER_EVENT` here?

Also should we have container events for:

- `handleResolved`
- `handleEvidenceSubmitted`
- `handleRuled`

And remove the ones we are not using from the list?

kanaagawa commented on packages/govern-subgraph/schema.graphql:
```diff
@@ -16,28 +16,18 @@ type Govern @entity {
   roles: [Role!]
 }
 
-type Execution @entity {
```
I think we need to have some way to track the execution that a Govern instance does so we keep a record of the results of the given execution.

Evalir replied:
We could probably just record these on the `data` field of the corresponding `ContainerEvent`?

bpierre commented on packages/govern-subgraph/schema.graphql:
```diff
@@ -16,28 +16,18 @@ type Govern @entity {
-  games: [OptimisticGame!] @derivedFrom(field: "executor")
-  executions: [Execution!]
+  correspondingEntries: [RegistryEntry!] @derivedFrom(field: "executor")
+  containers: [Container]!
```
```suggestion
  containers: [Container!]!
```

kanaagawa commented on packages/govern-subgraph/src/Govern.ts:
```diff
@@ -7,23 +7,17 @@ import {
-  const execution = loadOrCreateExecution(event)
+  const currentContainers = govern.containers
 
-  execution.results = event.params.execResults
```
This is the only place we can keep track of the results of the execution. Maybe have it as part of the `Execute` container event.

Evalir replied:
Yeah, definitely agree. :)

bpierre commented on packages/govern-subgraph/schema.graphql:
```diff
@@ -16,28 +16,18 @@ type Govern @entity {
+  config: Config!
+  correspondingEntries: [RegistryEntry!] @derivedFrom(field: "queue")
+  queued: [Container!]
+  challenged: [Challenge]!
```
I might be wrong, but I think we can use `[E!]!` most of the time (non-null arrays of non-null entries)?

```suggestion
  correspondingEntries: [RegistryEntry!]! @derivedFrom(field: "queue")
  queued: [Container!]!
  challenged: [Challenge!]!
```

kanaagawa commented on packages/govern-subgraph/src/GovernQueue.ts:
```diff
@@ -11,148 +11,182 @@ import {
-
-  const schedule = new ScheduleEntity(event.params.containerHash.toHexString())
-
-  const scheduleDeposit = loadOrCreateCollateral(event, '1')
```
We need to track the `Collateral` at the time the `Container` was schedule.

kanaagawa commented on packages/govern-subgraph/src/GovernQueue.ts:
```diff
@@ -11,148 +11,182 @@ import {
+  )
+  containerEvent.createdAt = event.block.timestamp
+  containerEvent.type = SCHEDULE_CONTAINER_EVENT
+  containerEvent.data = []
```
Maybe we can include the collateral as part of the data here.

0xGabi committed:
Subgraph: use ContainerEvents to handle data

0xGabi committed:
Subgraph: remove reference to challenged

kanaagawa commented on packages/govern-subgraph/src/Govern.ts:
```diff
@@ -5,17 +5,25 @@ import {
+    container,
+    event.block.timestamp,
+    EXECUTE_CONTAINER_EVENT,
+    [event.params.execResults.toString()]
```
I was not sure if we should instead return a data element per result here.

Evalir commented on packages/govern-subgraph/schema.graphql:
```diff
@@ -67,7 +67,7 @@ type ContainerEvent @entity {
   type: ContainerEventType
   createdAt: BigInt
-  data: [String]!
+  data: [String!]!
```
Btw, small thing, let's change this to bytes so we can encode anything here!

kanaagawa replied:
The main reason I didn't do it yet was because I'm not sure how to handle `BigInt`. What encoding do you suggest?

izqui committed:
Core, erc3k: Remove veto collateral (#98)

* Core, utils: Implement adaptative ERC165 (#87)

* Core, utils: Implement adaptitive ERC165

* Expose function

* Consistency

* Apply suggestions from code review

Co-authored-by: Enrique Ortiz <enrique@aragon.one>

* Apply suggestions from code review

Co-authored-by: Enrique Ortiz <enrique@aragon.one>

* ERC3k: hash missing payload fields (#89)

* Cleanup: core & create packages (#90)

* Cleanup: core & create repos

* Fix import

* Core, create, utils: Proxies in Factory (#88)

* Utils: Implement ERC1167ProxyFactory

* Core, utils: Implement Initializable

* Core, create, utils: use proxies in factory

* Fix CI, CD

* Fix test

* pump gas

Co-authored-by: Enrique Ortiz <hi@enriqueortiz.dev>

* chore: update main npm package readme

* Server: add pending review changes (#71)

* Core, erc3k: Remove veto collateral

* Make tests pass

* Fix subgraph

* Update to new registry

* Fix subgraphs from conflicts

Co-authored-by: Enrique Ortiz <enrique@aragon.one>
Co-authored-by: Enrique Ortiz <hi@enriqueortiz.dev>
Co-authored-by: Gabriel Garcia <gabrielpk.18@gmail.com>

Evalir committed:
Merge branch 'master' into subgrpah-rename

Evalir committed:
Fix CI reegistry address

bpierre committed:
Subgraph container events: use a shared interface (#101)

Add a ContainerEvent interface implemented by every container event entity.

Also:

- Add a finished property to the SubmitEvidence container event.
- Use a separate handler for each container event type.
- Remove build warnings (see #102).
- Update the subgraph to reflect the contract changes (see #103).
